src/lib/error/api-error.ts

```diff
@@ -0,0 +1,270 @@
+    | typeof UnleashApiErrorTypes[number]
+    | typeof UnleashApiErrorTypes2[number];
+
+const statusCode = (errorKind: UnleashApiErrorKind): number => {
+    switch (errorKind) {
+        case 'ContentTypeError':
+            return 415;
+        case 'ValidationError':
+            return 400;
+        case 'BadDataError':
+            return 400;
+        case 'BadRequestError':
+            return 400;
+        case 'OwaspValidationError':
+            return 400;
+        case 'PasswordUndefinedError':
+            return 400;
+        case 'MinimumOneEnvironmentError':
+            return 400;
+        case 'InvalidTokenError':
+            return 401;
+        case 'NoAccessError':
+            return 403;
+        case 'UsedTokenError':
+            return 403;
+        case 'InvalidOperationError':
+            return 403;
+        case 'IncompatibleProjectError':
+            return 403;
+        case 'OperationDeniedError':
+            return 403;
+        case 'NotFoundError':
+            return 404;
+        case 'NameExistsError':
+            return 409;
+        case 'FeatureHasTagError':
+            return 409;
+        case 'RoleInUseError':
+            return 400;
+        case 'ProjectWithoutOwnerError':
+            return 409;
+        case 'UnknownError':
+            return 500;
+        case 'InternalError':
+            return 500;
+        case 'PasswordMismatch':
+        case 'PasswordMismatchError':
+            return 401;
+        case 'DisabledError':
+            return 422;
+        case 'NotImplementedError':
+            return 405;
+        case 'NoAccessError':
+            return 403;
+        case 'AuthenticationRequired':
+            return 401;
+    }
+};
+
+type ValidationErrorDescription = { description: string; path?: string };
+
+type UnleashErrorData =
+    | {
+          message: string;
+          documentationLink?: string;
+      } & (
+          | {
+                name: Exclude<
+                    UnleashApiErrorKind,
+                    | 'NoAccessError'
+                    | 'AuthenticationRequired'
+                    | 'ValidationError'
+                    | 'BadDataError'
+                    | 'BadRequestError'
+                >;
+            }
+          | {
+                name: 'NoAccessError';
+                permission: string;
+            }
+          | {
+                name: 'AuthenticationRequired';
+                path: string;
+                type: string;
+            }
+          | {
+                name:
+                    | 'ValidationError'
+                    | 'BadDataError'
+                    | 'BadRequestError'
+                    | 'MinimumOneEnvironmentError';
+                errors: [
+                    ValidationErrorDescription,
+                    ...ValidationErrorDescription[],
+                ];
+            }
+      );
+
+export class UnleashError extends Error {
+    id: string;
+
+    name: UnleashApiErrorKind;
```

Comment:
In the bounded context (linguistic boundaries) of Unleash error I'd stick to either name or kind but not both. Since we're not in the bounded context of type theory I vote for name over kind.

Reply:
Sure! So would you call the type `UnleashApiErrorName`?

Reply:
Yeap

Reply:
Addressed in 6a16be1437cd929212acd8c95b303c7e2683bfcd